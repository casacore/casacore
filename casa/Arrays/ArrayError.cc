//# ArrayError.cc: Exception classes thrown by Array and related classes/functions
//# Copyright (C) 1993,1994,1995,1997,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

// .SUMMARY General, Indexing, and Conformace errors thrown by Array classes.

#include <casacore/casa/Arrays/ArrayError.h>


namespace casa { //# NAMESPACE CASA - BEGIN

ArrayError::ArrayError(Category c) : AipsError("ArrayError",c) {}

ArrayError::ArrayError(const Char *m,Category c) : AipsError(m,c) {}

ArrayError::ArrayError(const String &m,Category c) : AipsError(m,c) {}

ArrayError::~ArrayError() throw(){}



ArrayIndexError::ArrayIndexError(Category c) : ArrayError("ArrayIndexError",c) {}

ArrayIndexError::ArrayIndexError(const Char *m,Category c) : ArrayError(m,c) {}

ArrayIndexError::ArrayIndexError(const String &m,Category c) : ArrayError(m,c) {}

ArrayIndexError::ArrayIndexError(const IPosition &in, 
				 const IPosition &sh, const Char *m,Category c)
: ArrayError(m,c),
  i(in),
  l(sh)
{
    // Nothing
}

ArrayIndexError::~ArrayIndexError() throw(){}

IPosition ArrayIndexError::index() const
{
    return i;
}

IPosition ArrayIndexError::shape() const
{
    return l;
}



ArrayConformanceError::ArrayConformanceError(Category c) 
: ArrayError("ArrayConformanceError",c) 
{
    // Nothing
}

ArrayConformanceError::ArrayConformanceError(const Char *m,Category c) : ArrayError(m,c) {}

ArrayConformanceError::ArrayConformanceError(const String &m,Category c) : ArrayError(m,c) {}

ArrayConformanceError::~ArrayConformanceError() throw(){}



ArrayNDimError::ArrayNDimError(Int ndim1, Int ndim2, const Char *m,Category c)
: ArrayConformanceError(m + String(" ndim ") + String::toString(ndim1)
                        + " differs from " + String::toString(ndim2), c),
  r1(ndim1),
  r2(ndim2)
{}

ArrayNDimError::~ArrayNDimError() throw(){}

void ArrayNDimError::ndims(Int &ndim1, Int &ndim2) const
{
    ndim1 = r1; 
    ndim2 = r2;
}



ArrayShapeError::ArrayShapeError(const IPosition &s1, const IPosition & s2,
				 const Char *m,Category c)
: ArrayConformanceError(m + String(" shape ") + s1.toString()
                        + " differs from " + s2.toString(), c),
  sh1(s1), sh2(s2)
{
    // Nothing
}

ArrayShapeError::~ArrayShapeError() throw(){}

void ArrayShapeError::shapes(IPosition &shape1, IPosition &shape2) const
{
    shape1 = sh1; shape2 = sh2;
}



ArrayIteratorError::ArrayIteratorError(Category c) : ArrayError("ArrayIteratorError",c) {}

ArrayIteratorError::ArrayIteratorError(const Char *m,Category c) : ArrayError(m,c) {}

ArrayIteratorError::ArrayIteratorError(const String &m,Category c) : ArrayError(m,c) {}

ArrayIteratorError::~ArrayIteratorError() throw(){}



ArraySlicerError::ArraySlicerError(Category c) : ArrayError("Slicer error",c) {}

ArraySlicerError::ArraySlicerError(const String &m,Category c)
: ArrayError("Slicer error:" + m,c) {}

ArraySlicerError::~ArraySlicerError() throw(){}

} //# NAMESPACE CASA - END

