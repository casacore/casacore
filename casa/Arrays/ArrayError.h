//# ArrayError.h: Exception classes thrown by Array and related classes/functions
//# Copyright (C) 1993,1994,1995,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ARRAYERROR_H
#define CASA_ARRAYERROR_H

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/Exceptions/Error.h>
#include <casacore/casa/Arrays/IPosition.h>


namespace casacore { //# NAMESPACE CASACORE - BEGIN

// <summary> The base class for all Array exception classes. </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// ArrayError is the base class for all the Array-specific exception classes,
// i.e. if it is caught you will catch (through inheritance) all Array-specific
// exceptions. Note that (presently, anyway) the Array classes will throw
// a few non-Array exceptions.
// <srcblock>
// try {
//    // Some lines, functions, ...
// } catch (ArrayError x) {
//    // Array specific errors
// } catch (AipsError x) {
//    // All other errors caught here.
// }
// </srcblock>
//
//# There are too many Array related error classes. Some should be deleted.

class ArrayError : public AipsError
{
public:
    // Initialize with the message "ArrayError."
    ArrayError(Category c=GENERAL);
    // Initialize with the supplied message.
    ArrayError(const Char *m,Category c=GENERAL);
    // Initialize with the supplied message.
    ArrayError(const String &m,Category c=GENERAL);
    ~ArrayError() throw();
};


// <summary> An error thrown when an index is out of range </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// The ArrayIndexError class, which is derived from ArrayError, is intended
// to be thrown when an index is out-of-bounds. It contains within it
// the offending index, as well as the shape of the array which
// is being indexed. This should be multiply-derived from
// indexError<T> defined in Error.h.
class ArrayIndexError : public ArrayError
{
public:
    // Initialize with the message "ArrayIndexError".
    ArrayIndexError(Category c=BOUNDARY);
    // Initialize with the supplied message, the index and shape are null.
    ArrayIndexError(const Char *m,Category c=BOUNDARY);
    // Initialize with the supplied message, the index and shape are null.
    ArrayIndexError(const String &m,Category c=BOUNDARY);
    // Initialize with a given out-of-bounds index, as well as the shape
    // of the array and a supplied message.
    ArrayIndexError(const IPosition &index, const IPosition &shape, 
		    const Char *m="ArrayIndexError",Category c=BOUNDARY);
    ~ArrayIndexError() throw();
    // The out-of-bounds index.
    IPosition index() const;
    // The shape of the violated array.
    IPosition shape() const;
private:
    //# index, offset, length
    IPosition i,l;
};


// <summary> An error thrown when two arrays do not conform </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// The ArrayConformanceError class is the base class for all errors thrown
// because two arrays are not conformant. See also the ArrayShapeError and
// ArrayNDimError classes which are derived from it. This error, or one derived
// from it, os normally thrown from a binary operation (arithmetic, logical,
// assignment, etc).
class ArrayConformanceError : public ArrayError
{
public:
    // Initialize the message with "ArrayConformanceError".
    ArrayConformanceError(Category c=CONFORMANCE);
    // Initialize with a supplied message.
    ArrayConformanceError(const Char *m,Category c=CONFORMANCE);
    // Initialize with a supplied message.
    ArrayConformanceError(const String &m,Category c=CONFORMANCE);
    ~ArrayConformanceError() throw();
};


// <summary> Thrown when two arrays have different dimensionality </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// An ArrayNDimError is derived from ArrayConformanceError. It is thrown when
// two arrays are non-conformant by virtue of having different dimensionality.
// It holds within it the two dimensions.
class ArrayNDimError : public ArrayConformanceError
{
public:
    // Define the two (presumably different) messages and optionally
    // supply a message.
    ArrayNDimError(Int dim1, Int dim2, const Char *m="ArrayNDimError",Category c=CONFORMANCE);
    ~ArrayNDimError() throw();
    // Return the stored dimensions. NB modifies arguments.
    void ndims(Int &dim1, Int &dim2) const; // modifies arguments
private:
    Int r1, r2;
};


// <summary> An error thrown when two arrays have different shapes </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// An ArrayShapeError is derived from ArrayConformanceError. It is thrown when
// two arrays are non-conformant by virtue of having different shapes.
// It holds within it the two different two shapes.
class ArrayShapeError : public ArrayConformanceError
{
public:
    // Define an ArrayShapeError with the two (presumably different) shapes
    // and an optional supplied message.
    ArrayShapeError(const IPosition &shape1, const IPosition &shape2,
		     const Char *m="ArrayShapeError",Category c=CONFORMANCE);
    ~ArrayShapeError() throw();
    // Get back the stored shapes. NB modifies arguments.
    void shapes(IPosition &, IPosition &) const;  // modifies arguments
private:
    IPosition sh1, sh2;
};


// <summary> An error thrown by an ArrayIterator </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// An ArrayIteratorError is thrown by an array iterator or related class
// (e.g. VectorIterator).
class ArrayIteratorError : public ArrayError
{
public:    
    // Initialize with the message "ArrayIteratorError.
    ArrayIteratorError(Category c=BOUNDARY);
    // Initialize with the supplied message
    ArrayIteratorError(const Char *m,Category c=BOUNDARY);
    // Initialize with the supplied message
    ArrayIteratorError(const String &m,Category c=BOUNDARY);
    ~ArrayIteratorError() throw();
};


// <summary> An error thrown by an Slicer member function </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// An ArraySlicerError is thrown by an  Slicer member function.
class ArraySlicerError : public ArrayError
{
public:    
    // Initialize with the message "Slicer error."
    ArraySlicerError(Category c=GENERAL);
    // Initialize with ArraySlicerError plus the supplied message
    ArraySlicerError(const String &m,Category c=GENERAL);
    ~ArraySlicerError() throw();
};


} //# NAMESPACE CASACORE - END

#endif
