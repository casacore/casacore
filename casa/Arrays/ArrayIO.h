//# ArrayIO.h: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1997,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ARRAYIO_H
#define CASA_ARRAYIO_H

//# Includes
#include <casa/aips.h>
#include <casa/BasicSL/String.h>

//# Forward declarations
#include <casa/iosfwd.h>

namespace casa { //# NAMESPACE CASA - BEGIN

class AipsIO;
class LogIO;
class IPosition;
template<class T> class Array;
template<class T> class Matrix;
template<class T> class Vector;
template<class T> class Cube;
template<class T> class Block;

// <summary>
//    Input/output operators for Arrays.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Paul Shannon" date="1995/02/21" tests="" demos="">
// This header was reviewed and revised with the goal of making it an
// example for those writing global function header files.
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Array>Array</linkto>
//   <li> ostream
//   <li> <linkto class=AipsIO>AipsIO</linkto>
// </prerequisite>

// <etymology>
//  ArrayIO is simply the conventional shorthand for "array input/output".
// </etymology>

// <synopsis> 
// These global functions provide easy input and output of (possibly)
// large and (possibly) multi-dimensional arrays.  Iteration through
// entire arrays is done behind the scenes, with no effort required
// of the client programmer.  
// These functions are global, rather than member functions of the
// Array class, because of the well-known C++ requirement that the first
// argument to an operator function (as it is declared) is the 
// left operand when the function is called.
// </synopsis>


// <example>
// <srcblock>
// IPosition shape (3,10,10,3);
// Array <Float> array (shape);
// // ...initialize and manipulate the array...
// cout << "result: " << array;
// </srcblock>
//
// <motivation>
// Effortless input/output is clearly a big win.
// </motivation>
//
// <todo asof="1997/01/15">
// </todo>

// <linkfrom anchor="Array IO" classes="Array Vector Matrix Cube">
//    <here>Array IO</here> -- Input/output operators for Arrays.
// </linkfrom>
//
// <group name="Array IO">


// Write out an ascii representation of an array of any dimensionality.
// Arrays of dimensionality 3 or greater are written out vector by vector,
// preceeded by the position of the start of the vector. If the origin of
// the array isn't zero it is printed. The shape of the array is always
// printed.

template<class T> ostream &operator << (ostream &, const Array<T> &);

// Write a formatted copy of the array to the LogIO output object. Merely calls
// the ostream operator<< in turn.
template<class T> LogIO &operator<<(LogIO &os, const Array<T> &a);

// Read an ascii representation of an array. All types with an <src><<</src>
// operator can be handled. The basic format of the input should be:
// <srcblock>
//	[element element element ....] 
// </srcblock>
// Elements are separated by whitespace, or a comma, optionally surrounded
// by white space. <br>
// <note role=warning> Some input routines read fields between blank spaces. This
// is (at the moment) especially true for Quantities and Strings.
// In those cases
// the separator should be blank (or a comma following a blank), and the
// end ']' should have a blank in front.
// A crude fix for String arrays having separators <src>,</src> and <src>]</src>
// without blanks preceding has been made; but slows routines down </note>
// The default input is a vector of unspecified length. The input shape
// can be changed by pre-pending the input with:
// <srcblock>
//	{[shape]}
// </srcblock>
// where shape is an unsigned integer vector. The shape will be used to check
// the input length; and, depending on the possibility, to resize/reshape the
// result. However, reshaping of e.g. a Vector to a Matrix cannot be done, and
// the result will stay in the form asked.<br>
// Input order is row major, however by preceding the input with:
// <srcblock>
//	{T[shape]}	
// </srcblock>
// the order will be reversed.<br>
// Reshaping of the Array provided will depend on the type of Array and its
// state. If a general Array, the shape will be
// as defined by user. If fixed Array (e.g. Matrix, Vector, Cube) the number
// of dimesnsions will be kept. If the user specified more dimensions
// then supported (e.g. 3 for Matrix), the last dimesions will be collapsed.
// If less dimensions are specified, the missing ones will be set to 1.
// will be kept.<br>
// The read() version can be used to force a shape (ip), or an input
// transpose (it) (which can be undone by the user specifying transpose).
//
// <group>

template<class T> istream &operator >> (istream &s, Array<T> &x);
template<class T> Bool read(istream &s, Array<T> &x,
			    const IPosition *ip=0, Bool it=False);
// </group>

// General read support function for matrices.
// In principle these functions will not be
// used by general user, but could be. They can be used by Array type
// classes (like Slice, Lattice) to do the work of comparable input
// functions as the one for Arrays.
// In these functions p is the shape
// of the returned Block x. This shape is either deduced from the user
// specification; made equal to (1, nelements) if no user shape is
// given; is set to ip if specified. The function will return False (and
// p = (0)) in the case of an invalid input element; a number of elements
// input not equal to ip (if specified); the shape given by user as input
// does not conform to ip (if given) or the number of elements input.<br>
// trans will be True if transpose asked by user; or if forced by it.


template<class T> Bool readArrayBlock(istream &s, Bool &trans,
				      IPosition &p,
				      Block<T> &x,
				      const IPosition *ip=0, Bool it=False);


// Read or write a binary representation of an Array to a file. Very
// useful for saving arrays and restoring them later.
// <br>The putArray function is put in for forwards compatibility
// of images (so new images can be read with old release).
//
// <group>

template<class T> AipsIO &operator<< (AipsIO &, const Array<T> &);
template<class T> void putArray (AipsIO &, const Array<T> &, const Char* name);
template<class T> AipsIO &operator>> (AipsIO &, Array<T> &);

// </group>

// </group>




// <summary>
// Global functions to read/write binary arrays from/to a file.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="95Mar31" tests="" demos="">
// </reviewed>

// <synopsis>
// These global functions provide disk read/write functions for an Array of
// binary numbers.  The write operation is useful, for example, to dump an
// image in binary form to disk so that it can be displayed with an external
// utility such as SAOimage.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    write_array (picture, fileName);
// </srcblock>
// </example>

// <todo asof="">
//   <li> These functions should eventually be replaced with something
//        more sophisticated.
// </todo>

// <linkfrom anchor="Array binary IO" classes="Array Vector Matrix Cube">
//    <here>Array binary IO</here> -- Simple binary input/output for Arrays.
// </linkfrom>

// <group name=Array binary IO>

// Write the values of an array in binary format into a file with
// the given name.
// The values are stored in local format, thus are not converted
// to a canonical format as
// <linkto class="AipsIO:description">AipsIO</linkto>
// does.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void write_array (const Array<T>& the_array, const String& fileName);

template <class T>
inline void write_array (const Array<T>& the_array, const Char* fileName)
    { write_array (the_array, String(fileName)); }
// </group>

// Read the values of an array in binary format from a file with
// the given name.
// The number of values read is the size of the Array, thus the file
// should at least contain that number of values.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void read_array (Array<T>& the_array, const String& fileName);

template <class T>
inline void read_array (Array<T>& the_array, const Char* fileName)
    { read_array (the_array, String(fileName)); }
// </group>

// </group>



// <summary>
// Global functions for Matrix/Vector input/output using ASCII format.
// </summary>

// <use visibility=export>

// <prerequisite>
//   <li> <linkto class=Matrix>Matrix</linkto>
//   <li> <linkto class=Vector>Vector</linkto>
// </prerequisite>

// <synopsis>
// These global functions support file I/O between ASCII files and
// Matrices or Vectors.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    writeAsciiMatrix (picture, fileName);
// </srcblock>
// </example>

// <linkfrom anchor="Array Ascii IO" classes="Vector Matrix">
//    <here>Array Ascii IO</here> -- Simple Ascii input/output for Arrays.
// </linkfrom>

// <group name=Array Ascii IO>

// These routines read and write a Matrix of data.  The first line of
// input will be examined to determine the number of columns in the matrix.
// The maximum number of columns provided for is 100.  Each item may be up 
// to 50 characters long.
//
// Each item must be separated from others by one (or more) blank column.
// The "line" may be up to 1024 characters long.  Each subsequent line must
// contain the SAME number of items as the first line but may be any length
// (up to 1024 characters).
//
// The matrix need NOT be square.  
//
// The matrix should be declared but NOT dimensioned in the calling program.

// <group>
template <class T>
void readAsciiMatrix (Matrix<T>& mat, const Char* fileName);

template <class T>
void writeAsciiMatrix (const Matrix<T>& mat, const Char* fileName);
// </group>


// These two functions read and write a Vector of data.  The input
// may be arranged in any format (i.e. It may be recorded as one value per
// line or it may be recorded with all values on a single line).
// Values must be separated by whitespace.

// <group>
template <class T>
void readAsciiVector (Vector<T>& vec, const Char* fileName);

template <class T>
void writeAsciiVector (const Vector<T>& vec, const Char* fileName);
// </group>

// </group>



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Arrays/ArrayIO.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
