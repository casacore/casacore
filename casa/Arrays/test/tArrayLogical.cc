//# tArrayLogical.cc: Test program for Array logical operators
//# Copyright (C) 1993,1994,1995,1996,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# If AIPS_DEBUG is not set, the Assert's won't be called.
#if !defined(AIPS_DEBUG)
#define AIPS_DEBUG
#endif

//# For extra debugging
#if !defined(AIPS_ARRAY_INDEX_CHECK)
#define AIPS_ARRAY_INDEX_CHECK
#endif

#include <casacore/casa/iostream.h>

#include <casacore/casa/aips.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/BasicSL/Complex.h>
#include <casacore/casa/BasicMath/Math.h>
#include <casacore/casa/Utilities/Assert.h>

#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Arrays/ArrayMath.h>
#include <casacore/casa/Arrays/ArrayLogical.h>
#include <casacore/casa/Arrays/ArrayIO.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/casa/Arrays/Cube.h>
#include <casacore/casa/Arrays/ArrayError.h>
#include <casacore/casa/Arrays/LogiVector.h>


#include <casacore/casa/namespace.h>
int main()
{
    try {
        {
            cout << endl << "Testing Array logical operators." << endl;

            Vector<Int> x(5), y(5);
            LogicalVector b(5), c(5);
            x=1;
            indgen (y);

            cout << endl << "x= " << endl;
            cout << x << endl;
            cout << endl << "y= " << endl;
            cout << y << endl;

            AlwaysAssertExit (allSame(x));
            AlwaysAssertExit (!allSame(y));

            b = (x <= y);
            cout << endl << "b= (x <= y) = " << endl;
            cout << b << endl;

            b = (x < y);
            cout << endl << "b= (x < y) = " << endl;
            cout << b << endl;

            b = (x >= y);
            cout << endl << "b= (x >= y) = " << endl;
            cout << b << endl;

            b = (x > y);
            cout << endl << "b= (x > y) = " << endl;
            cout << b << endl;

            b = (x == y);
            cout << endl << "b= (x == y) = " << endl;
            cout << b << endl;

            b = (x != y);
            cout << endl << "b= (x != y) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y <= 1);
            cout << endl << "b= (y <= 1) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y < 1);
            cout << endl << "b= (y < 1) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y >= 1);
            cout << endl << "b= (y >= 1) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y > 1);
            cout << endl << "b= (y > 1) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y == 1);
            cout << endl << "b= (y == 1) = " << endl;
            cout << b << endl;

            b = ((const Array<Int> &)y != 1);
            cout << endl << "b= (y != 1) = " << endl;
            cout << b << endl;

            b = (1 <= (const Array<Int> &)y);
            cout << endl << "b= (1 <= y) = " << endl;
            cout << b << endl;

            b = (1 < (const Array<Int> &)y);
            cout << endl << "b= (1 < y) = " << endl;
            cout << b << endl;

            b = (1 >= (const Array<Int> &)y);
            cout << endl << "b= (1 >= y) = " << endl;
            cout << b << endl;

            b = (1 > (const Array<Int> &)y);
            cout << endl << "b= (1 > y) = " << endl;
            cout << b << endl;

            b = (1 == (const Array<Int> &)y);
            cout << endl << "b= (1 == y) = " << endl;
            cout << b << endl;

            b = (1 != (const Array<Int> &)y);
            cout << endl << "b= (1 != y) = " << endl;
            cout << b << endl;

            b = ! ((const Array<Int> &)y >= 3);
            c =   ((const Array<Int> &)y <  3);
            cout << endl << "b= ! (y >= 3) = " << endl;
            cout << b << endl;
            cout << "c=   (y <  3) = " << endl;
            cout << c << endl;
            cout << "allEQ (b, c) = " << allEQ (b, c) << endl;


            Vector<Double> x1(2), x2(2);
            x1 = 10000; x2(0) = 10001; x2(1) = 10002;
            cout << endl << " near(x1, x2, 0.99e-4) " << near(x1, x2, 0.99e-4);
            cout << endl << " near(x1, x2, 1.01e-4) " << near(x1, x2, 1.01e-4);
            cout << endl << " near(x1, x2, 2.01e-4) " << near(x1, x2, 2.01e-4);
            cout << endl << " nearAbs(x1, x2, 0.99) " << nearAbs(x1, x2, 0.99);
            cout << endl << " nearAbs(x1, x2, 1.01) " << nearAbs(x1, x2, 1.01);
            cout << endl << " nearAbs(x1, x2, 2.01) " << nearAbs(x1, x2, 2.01);

            cout << endl << "OK" << endl;
        }
    } catch (AipsError x) {
        cout << "\nCaught an exception: " << x.getMesg() << endl;
    } 

    cout << "OK" << endl;
    return 0;
}
