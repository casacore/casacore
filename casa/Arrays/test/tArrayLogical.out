
Testing Array logical operators.

x= 
[1, 1, 1, 1, 1]

y= 
[0, 1, 2, 3, 4]

b= (x <= y) = 
[0, 1, 1, 1, 1]

b= (x < y) = 
[0, 0, 1, 1, 1]

b= (x >= y) = 
[1, 1, 0, 0, 0]

b= (x > y) = 
[1, 0, 0, 0, 0]

b= (x == y) = 
[0, 1, 0, 0, 0]

b= (x != y) = 
[1, 0, 1, 1, 1]

b= (y <= 1) = 
[1, 1, 0, 0, 0]

b= (y < 1) = 
[1, 0, 0, 0, 0]

b= (y >= 1) = 
[0, 1, 1, 1, 1]

b= (y > 1) = 
[0, 0, 1, 1, 1]

b= (y == 1) = 
[0, 1, 0, 0, 0]

b= (y != 1) = 
[1, 0, 1, 1, 1]

b= (1 <= y) = 
[0, 1, 1, 1, 1]

b= (1 < y) = 
[0, 0, 1, 1, 1]

b= (1 >= y) = 
[1, 1, 0, 0, 0]

b= (1 > y) = 
[1, 0, 0, 0, 0]

b= (1 == y) = 
[0, 1, 0, 0, 0]

b= (1 != y) = 
[1, 0, 1, 1, 1]

b= ! (y >= 3) = 
[1, 1, 1, 0, 0]
c=   (y <  3) = 
[1, 1, 1, 0, 0]
allEQ (b, c) = 1

 near(x1, x2, 0.99e-4) [0, 0]
 near(x1, x2, 1.01e-4) [1, 0]
 near(x1, x2, 2.01e-4) [1, 1]
 nearAbs(x1, x2, 0.99) [0, 0]
 nearAbs(x1, x2, 1.01) [1, 0]
 nearAbs(x1, x2, 2.01) [1, 1]
OK
OK
